### PR TITLE
Downgrade const eval dangling ptr in final to future incompat lint

### PR DESCRIPTION
**Pull request by pnkfelix**
Short term band-aid for issue #121610, downgrading the prior hard error to a future-incompat lint (tracked in issue #122153).

Note we should not mark #121610 as resolved until after this (or something analogous) is beta backported.

### CODE REVIEW

**Comment by wesleywiser** [comment]
r? wesleywiser

**Comment by pnkfelix** [comment]
ugh I have to do more updating to placate our tidy+diagnostics checking

@rustbot author

**Comment by pnkfelix** [comment]
Okay hopefully the tidy and diagnostic checks are all placated now.

@rustbot reviewer

File: tests/ui/consts/miri_unleashed/mutable_references_err.rs
Comment (wesleywiser) [inline tests/ui/consts/miri_unleashed/mutable_references_err.rs:74]: It looks like we have some places where we were already generating an error but now we report the future incompatibility warning. That's perhaps a bit confusing but I don't think it needs to block the PR.

File: tests/ui/consts/miri_unleashed/mutable_references.stderr
Comment (riking) [inline tests/ui/consts/miri_unleashed/mutable_references.stderr:61]: missing HEX_DUMP replacement

**Comment by RalfJung** [comment]
The `compiler` part of the diff LGTM.

File: tests/ui/consts/miri_unleashed/mutable_references.stderr
Comment (RalfJung) [reply tests/ui/consts/miri_unleashed/mutable_references.stderr:61]: Ah, we probably want our usual normalization gadget in this file then
```
//@ normalize-stderr-test "(the raw bytes of the constant) \(size: [0-9]*, align: [0-9]*\)" -> "$1 (size: $$SIZE, align: $$ALIGN)"
//@ normalize-stderr-test "([0-9a-f][0-9a-f] |╾─*ALLOC[0-9]+(\+[a-z0-9]+)?(<imm>)?─*╼ )+ *│.*" -> "HEX_DUMP"
```

File: tests/ui/consts/miri_unleashed/mutable_references_err.rs
Comment (RalfJung) [reply tests/ui/consts/miri_unleashed/mutable_references_err.rs:74]: Note that this file uses `-Zunleash-the-miri-inside-of-you`. I am not sure if this can even be triggered on stable.

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (RalfJung) [inline compiler/rustc_lint_defs/src/builtin.rs:2794]: > failed to test example in lint docs for `const_eval_mutable_ptr_in_final_value` in /checkout/compiler/rustc_lint_defs/src/builtin.rs:2769: lint docs should contain the line `### Explanation`

File: tests/ui/consts/miri_unleashed/mutable_references_err.rs
Comment (oli-obk) [reply tests/ui/consts/miri_unleashed/mutable_references_err.rs:74]: Could just not emit the future incompat lint if unleash mode is on

Comment (RalfJung) [reply tests/ui/consts/miri_unleashed/mutable_references_err.rs:74]: We want the error in unleash mode, it's the only way unleash mode can be sound.

Comment (pnkfelix) [reply tests/ui/consts/miri_unleashed/mutable_references_err.rs:74]: > Could just not emit the future incompat lint if unleash mode is on

As a variation on the above, I was musing that if *any* future-incompat lint is set to deny or forbid, then we could stop emitting that boilerplate about the thing being potentially rejected in a future version of the compiler, just to cut down on the diagnostic flooding...

File: tests/ui/consts/miri_unleashed/mutable_references.stderr
Comment (pnkfelix) [reply tests/ui/consts/miri_unleashed/mutable_references.stderr]: Added in 03bdb0f0176891e196ab59a93cfc4e104a7273ac

File: compiler/rustc_lint_defs/src/builtin.rs
Comment (wesleywiser) [inline compiler/rustc_lint_defs/src/builtin.rs:2790]:
```suggestion
    ///
    /// {{produces}}
    ///
```

**Review by wesleywiser — APPROVED** [review]
r=me when CI is green 👍

**Comment by pnkfelix** [comment]
Regarding the beta-nomination: This ended up requiring a bit more changes than I had anticipated when we discussed it last week.

Another option, that may be more palatable for backporting, would be to just disable the hard-error with *no lint at all* on beta, and let the future-incompat code ride the train.

**Comment by pnkfelix** [comment]
oh. of course. need to handle tests that are on 32-bit targets. (sigh)

**Comment by wesleywiser** [comment]
> Almost all the changes are in tests. So in terms of the diffstat I don't think that would be much smaller?

I agree with Ralf, the functional changes are only ~60 lines with the rest being in test output and the added test case. I don't see an issue with backporting that.

**Comment by wesleywiser** [comment]
Thanks for getting this done @pnkfelix!

@bors r+ p=1

Resolves P-high issue we want to backport

**Comment by apiraino** [comment]
Beta backport approved as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-03-14/near/426551945). A backport PR will be authored by the release team at the end of the current development cycle.

@rustbot label +beta-accepted

**Pull request merged**
